server/src/test/java/org/elasticsearch/index/engine/LiveVersionMapTests.java

```diff
@@ -396,6 +397,32 @@ public void testPruneTombstonesWhileLocked() throws InterruptedException, IOExce
         assertEquals(0, map.getAllTombstones().size());
     }
 
+    public void testNeverLeaveStaleDeleteTombstone() throws Exception {
+        LiveVersionMap versionMap = new LiveVersionMap();
+        BytesRef uid = uid("1");
+        long versions = between(10, 1000);
+        for (long version = 1; version <= versions; version++) {
+            if (randomBoolean()) {
+                versionMap.beforeRefresh();
+                versionMap.afterRefresh(randomBoolean());
+            }
+            if (randomBoolean()) {
+                versionMap.enforceSafeAccess();
+            }
+            try (Releasable ignore = versionMap.acquireLock(uid)) {
+                if (randomBoolean()) {
+                    versionMap.putDeleteUnderLock(uid, new DeleteVersionValue(version, 1, 1, 1));
+                } else {
+                    versionMap.maybePutIndexUnderLock(uid, new IndexVersionValue(randomTranslogLocation(), version, 1, 1));
+                }
+                VersionValue storedValue = versionMap.getUnderLock(uid);
```

Comment:
can we also check that this return value correlates with what we expect? i.e., if the doc was just deleted it should be deleted, if it was indexed it should never be marked as deleted? (and found if we're in safe mode). Also can you check this property holds also not directly after indexing/deleting? (i.e., after refresh, enforceSafe access etc). Said differently - maybe also randomly do nothing under lock?